pkg/patterns/declarative/pkg/applier/direct_test.go

```diff
@@ -0,0 +1,194 @@
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+type directApplierTestSite struct {
+	Error     error
+	applyOpts *apply.ApplyOptions
+}
+
+func (d *directApplierTestSite) Run(a *apply.ApplyOptions) error {
+	d.applyOpts = a
+	return nil
+}
+
+// same manner with a function newUnstructuredDefaultBuilderWith on k8s.io/cli-runtime/pkg/resource/builder_test.go
+func (d *directApplierTestSite) NewBuilder(opt ApplierOptions) *resource.Builder {
+	return resource.NewFakeBuilder(
+		func(version schema.GroupVersion) (resource.RESTClient, error) {
+			return &fake.RESTClient{}, nil
+		},
+		func() (meta.RESTMapper, error) {
+			return testrestmapper.TestOnlyStaticRESTMapper(scheme.Scheme), nil
+		},
+		func() (restmapper.CategoryExpander, error) {
+			return resource.FakeCategoryExpander, nil
+		})
+}
+
+func newDirectApplierTest(d *directApplierTestSite) Applier {
+	return &DirectApplier{inner: d}
+}
+
+func TestDirectApply(t *testing.T) {
+	tests := []struct {
+		name               string
+		namespace          string
+		manifest           string
+		validate           bool
+		args               []string
+		err                error
+		expectApplyOptions *apply.ApplyOptions
+		expectCheckFunc    func(opt *apply.ApplyOptions) error
+	}{
+		{
+			name:      "manifest",
+			namespace: "",
+			manifest: `---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-operator
+  namespace: kube-system`,
+			expectApplyOptions: &apply.ApplyOptions{},
+			expectCheckFunc: func(opt *apply.ApplyOptions) error {
+				return nil
+			},
+		},
+		{
+			name:      "manifest with namespace",
+			namespace: "test-namespace",
+			manifest: `---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-operator
+  namespace: kube-system`,
+			expectApplyOptions: &apply.ApplyOptions{
+				Namespace: "test-namespace",
+			},
+			expectCheckFunc: func(opt *apply.ApplyOptions) error {
+				if opt.Namespace == "test-namespace" {
+					// validator is set, not nil.
+					return nil
+				} else {
+					return fmt.Errorf("namespace doesn't match to \"test-namespace\"")
+				}
+			},
+		},
+		{
+			name:      "manifest with validate",
+			namespace: "",
+			manifest: `---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-operator
+  namespace: kube-system`,
+			validate: true,
+			args:     []string{},
+			expectCheckFunc: func(opt *apply.ApplyOptions) error {
+				if opt.Validator != nil {
+					// validator is set, not nil.
+					return nil
+				} else {
+					return fmt.Errorf("validator is not nil")
+				}
+			},
+		},
+		{
+			name:      "manifest with prune",
+			namespace: "",
+			manifest: `---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: foo-operator
+  namespace: kube-system`,
+			args: []string{"--prune"},
+			expectApplyOptions: &apply.ApplyOptions{
+				Prune: true,
+			},
+			expectCheckFunc: func(opt *apply.ApplyOptions) error {
+				if opt.Prune != true {
+					return fmt.Errorf("prune is not set")
+				}
+				return nil
+			},
+		},
+		//		{
```

Comment:
I found that current codebase [doesn't refer `--prune-whitelist`](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/2072cf081b2e512b687c6bf152466646df1ac2c9/pkg/patterns/declarative/pkg/applier/direct.go#L92-L102).
That's why I commented out here.

Reply:
Do you use prune-whitelist?  If not, I propose we make the direct applier strict, i.e. it should reject unknown options.  Thus it should error if someone passes `prune-whitelist`.

If you do use it, can we map it to the underlying library?